src/Host/Client/Impl/BrokerServices/WebServer.cs

```diff
@@ -0,0 +1,155 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Common.Core;
+using Newtonsoft.Json;
+
+namespace Microsoft.R.Host.Client.BrokerServices {
+    public class WebServer {
+        private readonly string _newHost;
+        public string Host => _newHost;
+
+        private readonly int _newPort;
+        public int Port => _newPort;
+
+        private readonly string _remoteHost;
+        public string RemoteHost => _remoteHost;
+
+        private readonly int _remotePort;
+        public int RemotePort => _remotePort;
+
+        private HttpListener _listener;
+
+        private static List<WebServer> _servers = new List<WebServer>();
+        public static IReadOnlyList<WebServer> Servers => _servers;
+
+        private readonly RemoteUriWebService _service;
+        private IRemoteUriWebService RemoteUriService => _service;
+
+        private WebServer(string remoteHostIp, int remotePort, HttpClient httpClient) {
+            _newHost = IPAddress.Loopback.ToString();
+            _remoteHost = remoteHostIp;
+            _remotePort = remotePort;
+            _service = new RemoteUriWebService(httpClient);
+            Random r = new Random();
+
+            // if remote port is between 10000 and 32000, select a port in the same range.
```

Comment:
Do we actually care about ports being in the same range? It's not going to be the same port anyway.